scipy/signal/signaltools.py

```diff
@@ -2438,18 +2492,52 @@ def resample_poly(x, up, down, axis=0, window=('kaiser', 5.0)):
     n_post_pad = 0
     n_pre_remove = (half_len + n_pre_pad) // down
     # We should rarely need to do this given our filter lengths...
-    while _output_len(len(h) + n_pre_pad + n_post_pad, x.shape[axis],
+    while _output_len(len(h) + n_pre_pad + n_post_pad, n_in,
                       up, down) < n_out + n_pre_remove:
         n_post_pad += 1
     h = np.concatenate((np.zeros(n_pre_pad, dtype=h.dtype), h,
                         np.zeros(n_post_pad, dtype=h.dtype)))
     n_pre_remove_end = n_pre_remove + n_out
 
+    # Remove background depending on the padtype option
+    funcs = {'mean': np.mean, 'median': np.median,
+             'minimum': np.amin, 'maximum': np.amax}
+    if padtype == 'constant':
+        background_line = cval
+    elif padtype in funcs:
+        background_line = [funcs[padtype](x, axis=axis), 0]
+    elif padtype == 'line':
+        background_line = [x.take(0, axis),
+                           (x.take(-1, axis) - x.take(0, axis))*n_in/(n_in-1)]
+    else:
+        raise ValueError('padtype must be line, maximum, mean, median, minimum or constant')
+
+    if padtype == 'line' or padtype in funcs:
+        rel_len = np.linspace(0.0, 1.0, n_in, endpoint=False)
+        rel_len_nd = _reshape_nd(rel_len, x.ndim, axis)
+        background_in = np.expand_dims(background_line[0], axis) +\
+            np.expand_dims(background_line[1], axis) * rel_len_nd
+        x = x - background_in.astype(x.dtype)
```

Comment:
Just to make sure I am not misunderstanding: this linear case basically works as in the 1D example below, but with a trend line fit for each "line" to be filtered? I think that seems reasonable and is a nice way to not need extension outside the original bounds of the signal.

```Python
import numpy as np
from scipy.misc import electrocardiogram
from matplotlib import pyplot as plt
ecg = electrocardiogram()[:2000]
fig, axes = plt.subplots(2, 1, figsize=(6, 8))
axes[0].plot(ecg)
axes[0].set_title("original signal")
axes[1].plot(ecg - np.linspace(ecg[0], ecg[-1], ecg.size))
axes[1].set_title("linear trend removed")
plt.tightlayout()
```
![ecg_detrend](https://user-images.githubusercontent.com/6528957/62378339-9b3cc280-b512-11e9-8d0c-9c2c44fd68e8.png)

Reply:
Exactly! I first find the line joining the first and last elements and then remove it from the signal. After resampling, I add it back.

This is different to np.pad's edge (which is just constant padding with each edge value).